STYLEGUIDE.md

```diff
@@ -0,0 +1,716 @@
+const foo = `foo ${bar} bar`;
+
+/* Grey area: If template literals become too complex, feel free to concatenate with +. */
+const foo = `foo ${fn(bar +'bar')}`;
+```
+
+## Arrays
+
+### Use `.forEach` rather than `for`
+Use the functional language concepts of JavaScript in favor of the imperative ones, when possible.
+Use the correct function for your use case.
+In rare cases, when performance is of the essence, it makes sense to resort to the imperative `for` instead.
+
+❌ Examples of **incorrect** code:
+```ts
+/* Don't use for ... of if you can use .forEach */
+for (const item of array) {
+  ...
+}
+
+/* Don't use .forEach if you can use .map */
+const result = [];
+array.forEach(item => {
+  result.push(item.property);
+});
+```
+
+✅ Examples of **correct** code:
+```ts
+/* Use .forEach if possible */
+array.forEach(item => {
+  ...
+});
+
+/* Use .map, if appropriate */
+const result = array.map(item => item.property);
+```
+
+
+## `null` and `undefined`
+
+### Use optional parameters for optional values
+Use optional parameters, if there is a default value or if it semantically makes sense to omit the parameters when consuming the API.
+
+❌ Examples of **incorrect** code:
+```ts
+/* Don't use explicit undefined if the parameter is semantically optional */
+function foo(options: Options | undefined) {
+  ...
+}
+```
+
+✅ Examples of **correct** code:
+```ts
+/* Use optional parameter if the parameter is semantically optional */
+function foo(options?: Options) {
+  ...
+}
+
+/* Use default parameter if there is a default (making the parameter inherently optional for the consumer) */
+function foo(options: Options = {}) {
+  ...
+}
+```
+
+### Use `undefined` for explicit absence of values
+If a value semantically is not optional prefer using `undefined` for values that do not exist or are not defined.
+
+❌ Examples of **incorrect** code:
+```ts
+/* Don't use null for non-existent values */
+function foo(obj: SomeType | null) {
+  ...
+}
+```
+
+✅ Examples of **correct** code:
+```ts
+/* Use undefined for non-existent values */
+function foo(obj: SomeType | undefined) {
+  ...
+}
+```
+
+### Use `null` for intentional explicit absence of values or if it is part of an external API
+If a value can both be non-existent and intentionally absent and those states are semantically different it may make sense to allow the usage of `null`.
+Those cases should however be rather rare and are more common when calling external APIs, e. g. through HTTP requests.
+Consider whether there is a better API design that might avoid this.
+
+❌ Examples of **incorrect** code:
+```ts
+/* Don't use null for absence in general */
+function foo(obj?: SomeType | null) {
+  if (obj === null || obj === undefined) {
+    ...
+  }
+}
+```
+
+✅ Examples of **correct** code:
+```ts
+/* Use null if it semantically makes sense and is different from undefined */
+function foo(obj: SomeType | null | undefined) {
+  if (obj === null) {
+    ...
+  } else if (obj === undefined) {
```

Comment:
Shouldn't this be `typeof obj === 'undefined'`?

Reply:
Excellent question. I always did (and still do this), but I recently thought to myself, why should we?
The reason we were doing this is, because you could potentially reassign undefined as this is not a reserved keyword. I think, we should just not allow using undefined as a variable name (and I think this rule would allow us to do this: https://eslint.org/docs/rules/id-denylist).

The second reason (that I just found out to be honest), is that the typeof check also works for undeclared variables. I believe that in TypeScript and JS strict mode this can never happen, right?

Therefore I don't see a good reason to make it more complex / more verbose.

What do you think? Am I missing anything?

Reply:
I just remembered that this used to be a problem in the past but I just looked it up and since ES5 you can't assign a value to `undefined` and accessing an non-existant global is not a problem in typescript so I think we can safely use `obj === undefined`. 👍